### PR TITLE
websockets infrastructure

### PR DESCRIPTION
**Pull request by fridgerator**
### Requirements

Initial implementation of a websocket handler, manager and server.  

### Description of the Change

This PR allows for a websocket handler to be created within the application config block.  A user can also define their own `Socket` and `Channel` functionality (authentication, message handling, etc).  Example code below.

### Why Should This Be In Core?

* Allows for websocket communication to be easily added to any application.

### Features still TODO

* Heartbeat - server: to kill dead connections, client: to ensure connection to server is still active
* PubSub implementation through redis / postgres - for application scalability
* Allow for different decoding methods - allow the user to use something other that JSON if they want.
* Message buffering?
* Long polling fallback?

### Example Code

1. Create a socket and channel:
```crystal
struct UserSocket < Amber::WebSockets::ClientSocket
  channel "chat_room:*", ChatChannel

  def on_connect
    # This runs when the socket connects.  Authentication should happen here.
    # If `on_connect` returns false, the socket will be closed.
    true
  end
end

class ChatChannel < Amber::WebSockets::Channel
  def handle_joined
    # Called when a user joins a channel
  end

  def handle_message(msg)
    # Called when a message is sent to the channel
    # use rebroadcast! to broadcast to the message to all users subscribed to the channel
    rebroadcast!(msg)
  end
end
```

2. In the application config block, define the handler endpoint
```crystal
MyApp.config do
  ...
  socket_endpoint "/chat", UserSocket
 ...
end
```

3. Javascript library usage:
```javascript
let socket = new Amber.Socket("/chat");
socket.connect();

let channel = socket.channel('chat_room:123')
channel.join()

channel.on('msg:new', (msg) => { console.log('new message : ', msg) } )

channel.push('msg:new', {message: 'hey guys'})
```

### CODE REVIEW

File: src/amber/router/pipe/router.cr [hunk 1]
```diff
@@ -23,6 +24,11 @@ module Amber
         context
       end
 
+      def get_socket_handler(request)
+        raise Exceptions::RouteNotFound.new(request) if !socket_route_defined?(request)
```
Comment (eliasjpr) [inline src/amber/router/pipe/router.cr]: `unless socket_route_defined?(request)` maybe?

File: src/amber/websockets/client_socket.cr [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      def self.get_topic_channel(topic)
+        topic_channels = @@channels.select { |ch| ch[:path].split(":")[0] == topic }
+        return topic_channels[0][:channel] if topic_channels.any?
+      end
+
+      def initialize(@socket)
+        @id = @socket.object_id
+        @subscription_manager = SubscriptionManager.new
+        @socket.on_pong do |msg|
+          # TODO: setup heartbeat
```
Comment (eliasjpr) [inline src/amber/websockets/client_socket.cr]: Should we aadd an issue/card/story to keep track of this?

File: assets/js/websockets/socket.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    return channel
+  }
+
+  handleMessage (msg) {
+    msg = JSON.parse(msg.data)
+    this.channels.forEach((channel) => {
+      if (channel.topic === msg.topic) channel.handleMessage(msg)
+    })
+  }
+}
```
Comment (eliasjpr) [inline assets/js/websockets/socket.js]: New line

File: assets/js/websockets/channel.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  on (subject, callback) {
+    this.onMessageHandlers.push({subject: subject, callback: callback})
+  }
+
+  push (subject, payload) {
+    this.socket.ws.send(JSON.stringify({event: EVENTS.message, topic: this.topic, subject: subject, payload: payload}))
+  }
+}
```
Comment (eliasjpr) [inline assets/js/websockets/channel.js]: New line

**Comment by fridgerator** [comment]
Sorry i've been busy lately, I should have time to update this tonight.

**Pull request closed without merging**